### PR TITLE
ITSASU-775, added Agent TaskList controller and view

### PR DESCRIPTION
**Pull request by WozCapGem**

### CODE REVIEW

File: app/views/agent/AgentTaskList.scala.html [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    }}
+</ul>
+}
+
+
+@mainTemplate(title = messages("agent.business.task-list.title")) {
+
+    <h1 class="govuk-heading-xl">@messages("agent.business.task-list.title")</h1>
+    <p class="govuk-body-l govuk-!-font-weight-bold">
+        @clientName @clientNino
```
Comment (sb-dev) [inline app/views/agent/AgentTaskList.scala.html:143]: There is a line break between `clientName` and `clientNino` in the prototype.

File: app/views/agent/AgentTaskList.scala.html [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    accountingPeriodService: AccountingPeriodService
+)
+
+
+@(postAction: Call, viewModel: TaskListModel, clientName: String, clientNino: String)(implicit request: Request[_], messages: Messages, appConfig: AppConfig)
+
+
+@signUpSectionContent = @{
+    if(viewModel.taskListComplete ) {
+            continueButton()
```
Comment (sb-dev) [inline app/views/agent/AgentTaskList.scala.html:36]: On the prototype, the button's text is `Submit and continue`

File: it/controllers/agent/TaskListControllerISpec.scala [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+          UsersGroupsSearchStub.stubGetUsersForGroups(testGroupId)(NON_AUTHORITATIVE_INFORMATION, UsersGroupsSearchStub.successfulResponseBody)
+          EnrolmentStoreProxyStub.stubUpsertEnrolment(testMtdId, testNino)(NO_CONTENT)
+          EnrolmentStoreProxyStub.stubAllocateEnrolmentWithoutKnownFacts(testMtdId, testGroupId, testCredentialId)(CREATED)
+          EnrolmentStoreProxyStub.stubAssignEnrolment(testMtdId, testCredentialId)(CREATED)
+          EnrolmentStoreProxyStub.stubAssignEnrolment(testMtdId, testCredentialId2)(CREATED)
+
+          When("I call POST /task-list")
+          val res = IncomeTaxSubscriptionFrontend.submitTaskList()
+
+          Then("The result should have a status of SEE_OTHER and redirect to the confirmation page")
```
Comment (sb-dev) [inline it/controllers/agent/TaskListControllerISpec.scala:653]: `Then("The result should have a status of SEE_OTHER and redirect to the confirmation page")` needs updating.

File: it/controllers/agent/TaskListControllerISpec.scala [hunk 4]
```diff
@@ -0,0 +1,686 @@
...
+        IncomeTaxSubscriptionConnectorStub.stubGetSubscriptionDetails(BusinessAccountingMethod, NO_CONTENT)
+        IncomeTaxSubscriptionConnectorStub.stubGetSubscriptionDetails(Property, NO_CONTENT)
+        IncomeTaxSubscriptionConnectorStub.stubGetSubscriptionDetails(OverseasProperty, NO_CONTENT)
+
+        val serviceNameGovUk = " - Use software to report your client’s Income Tax - GOV.UK"
+
+        When("GET /business/task-list is called")
+        val res = IncomeTaxSubscriptionFrontend.getTaskList()
+
+        Then("Should return OK with the task list page")
```
Comment (sb-dev) [inline it/controllers/agent/TaskListControllerISpec.scala:678]: `Then("Should return OK with the task list page")` needs updating.

File: it/controllers/agent/TaskListControllerISpec.scala [hunk 5]
```diff
@@ -0,0 +1,686 @@
...
+          UsersGroupsSearchStub.stubGetUsersForGroups(testGroupId)(NON_AUTHORITATIVE_INFORMATION, UsersGroupsSearchStub.successfulResponseBody)
+          EnrolmentStoreProxyStub.stubUpsertEnrolment(testMtdId, testNino)(NO_CONTENT)
+          EnrolmentStoreProxyStub.stubAllocateEnrolmentWithoutKnownFacts(testMtdId, testGroupId, testCredentialId)(CREATED)
+          EnrolmentStoreProxyStub.stubAssignEnrolment(testMtdId, testCredentialId)(CREATED)
+          EnrolmentStoreProxyStub.stubAssignEnrolment(testMtdId, testCredentialId2)(CREATED)
+
+          When("I call POST /task-list")
+          val res = IncomeTaxSubscriptionFrontend.submitTaskList()
+
+          Then("The result should have a status of SEE_OTHER and redirect to the confirmation page")
```
Comment (sb-dev) [inline it/controllers/agent/TaskListControllerISpec.scala:608]: Then("The result should have a status of SEE_OTHER and redirect to the confirmation page") needs updating.

File: test/services/agent/SubscriptionOrchestrationServiceSpec.scala [hunk 6]
```diff
@@ -129,4 +130,48 @@ class SubscriptionOrchestrationServiceSpec extends MockSubscriptionService with
...
+        whenReady(res)(_ mustBe testSignUpIncomeSourcesFailure)
+      }
+
+      "create income sources from task list request fail and returns an error" in {
+        mockSignUpIncomeSourcesSuccess(testNino)
+        mockCreateIncomeSourcesFromTaskListFailure(testMTDID, testCreateIncomeSources)
+
+        whenReady(res)(_ mustBe testCreateSubscriptionFromTaskListFailure)
+      }
+
```
Comment (sb-dev) [inline test/services/agent/SubscriptionOrchestrationServiceSpec.scala:174]: Do we need a test for when `confirmAgentEnrollmentToSps` in `SubscriptionOrchestrationService` fails?

File: app/views/agent/AgentTaskList.scala.html (hunk 1 above)
Comment (WozCapGem) [reply app/views/agent/AgentTaskList.scala.html:143]: We were given a screenshot from Jitesh to work from which had both values on the same line

File: app/views/agent/AgentTaskList.scala.html [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+}
+
+@businesses = @{
+
+<ul class="app-task-list__items">
+    {viewModel.selfEmployments.map(se =>
+    <li class="app-task-list__item">
+            <span class="app-task-list__task-name">
+            <a href={s"${appConfig.incomeTaxSelfEmploymentsFrontendBusinessCheckYourAnswersUrl}?id=${se.id}&isEditMode=true"} aria-describedby={"section_2_item_1_status"+viewModel.selfEmployments.indexOf(se)}>{se.businessName.map(bn => bn.businessName).getOrElse("")} {se.businessTradeName.map(btn => btn.businessTradeName).getOrElse("")}</a>
```
Comment (sb-dev) [inline app/views/agent/AgentTaskList.scala.html:100]: `appConfig.incomeTaxSelfEmploymentsFrontendBusinessCheckYourAnswersUrl` is the individual link; you can create a new one for the Agent Sole Trader CYA pages.

File: app/views/agent/AgentTaskList.scala.html (hunk 1 above)
Comment (sb-dev) [reply app/views/agent/AgentTaskList.scala.html:143]:
```
  override val incomeTaxSelfEmploymentsFrontendClientBusinessCheckYourAnswersUrl: String = {
    s"${config.getString("income-tax-subscription-self-employed-frontend.url")}/report-quarterly/income-and-expenses/sign-up/self-employments/client/details/business-check-your-answers"
  }
```

File: app/views/agent/AgentTaskList.scala.html (hunk 7 above)
Comment (sb-dev) [reply app/views/agent/AgentTaskList.scala.html:100]: *app/config/frontendAppConfig.scala*
```
  override val incomeTaxSelfEmploymentsFrontendClientBusinessCheckYourAnswersUrl: String = {
    s"${config.getString("income-tax-subscription-self-employed-frontend.url")}/report-quarterly/income-and-expenses/sign-up/self-employments/client/details/business-check-your-answers"
  }
```

**Pull request merged**